File: cpp/tests/data_structure/test_segment_tree.cpp
#include "test_utility.hpp"
#include "reachset/data_structure/segment_tree.hpp"

using namespace reach;

TEST_SUITE("TestSegmentTree") {
TEST_CASE("CounterSegmentTree") {
    auto tree = CounterSegmentTree(0, 20);

    SUBCASE("activate joint segments") {
        tree.activate(0, 10);
        tree.activate(5, 15);

        auto vec_nodes_active = tree.get_active_nodes();
        vector<CounterTreeNode> vec_nodes_unwrapped;
        vec_nodes_unwrapped.reserve(vec_nodes_active.size());
        for (auto const& node:vec_nodes_active)
            vec_nodes_unwrapped.emplace_back(*node);

        for (auto const& node_active:{CounterTreeNode(0, 10), CounterTreeNode(10, 15)}) {
            CHECK(counter_node_in_nodes(node_active, vec_nodes_unwrapped));
        }
    }

    SUBCASE("activate disjoint segments") {
        tree.activate(0, 5);
        tree.activate(15, 20);

        auto vec_nodes_active = tree.get_active_nodes();
        vector<CounterTreeNode> vec_nodes_unwrapped;
        vec_nodes_unwrapped.reserve(vec_nodes_active.size());
        for (auto const& node:vec_nodes_active)
            vec_nodes_unwrapped.emplace_back(*node);

        for (auto const& node_active:{CounterTreeNode(0, 5), CounterTreeNode(15, 20)}) {
            CHECK(counter_node_in_nodes(node_active, vec_nodes_unwrapped));
        }

        tree.activate(5, 15);

        vec_nodes_active = tree.get_active_nodes();
        vec_nodes_unwrapped.clear();
        vec_nodes_unwrapped.reserve(vec_nodes_active.size());
        for (auto const& node:vec_nodes_active)
            vec_nodes_unwrapped.emplace_back(*node);

        for (auto const& node_active:{CounterTreeNode(0, 20)}) {
            CHECK(counter_node_in_nodes(node_active, vec_nodes_unwrapped));
        }

    }

    SUBCASE("deactivate after activate returns empty active nodes") {
        tree.activate(0, 10);
        tree.activate(5, 15);
        tree.deactivate(5, 15);
        tree.deactivate(0, 10);

        CHECK(tree.get_active_nodes().empty());

        tree.activate(0, 10);
        tree.activate(5, 15);
        tree.deactivate(0, 10);
        tree.deactivate(5, 15);

        CHECK(tree.get_active_nodes().empty());;

    }
}
}

File: tests/python/data_structure/test_segment_tree.py
from commonroad_reach.data_structure.segment_tree import CounterSegmentTree, CounterTreeNode, ToggleSegmentTree


def test_activate_joint_segments():
    tree = CounterSegmentTree(0, 20)
    tree.activate(0, 10)
    tree.activate(5, 15)

    list_nodes_active = tree.get_active_nodes()
    for node_active in [CounterTreeNode(0, 10), CounterTreeNode(10, 15)]:
        assert node_active in list_nodes_active


def test_activate_disjoint_segments():
    tree = CounterSegmentTree(0, 20)
    tree.activate(0, 5)
    tree.activate(15, 20)

    list_nodes_active = tree.get_active_nodes()
    for node_active in [CounterTreeNode(0, 5), CounterTreeNode(15, 20)]:
        assert node_active in list_nodes_active

    tree.activate(5, 15)
    list_nodes_active = tree.get_active_nodes()
    for node_active in [CounterTreeNode(0, 20)]:
        assert node_active in list_nodes_active


def test_deactivate_after_activate_returns_empty_active_nodes():
    tree = CounterSegmentTree(0, 20)

    tree.activate(0, 10)
    tree.activate(5, 15)
    tree.deactivate(5, 15)
    tree.deactivate(0, 10)

    assert len(tree.get_active_nodes()) == 0

    tree.activate(0, 10)
    tree.activate(5, 15)
    tree.deactivate(0, 10)
    tree.deactivate(5, 15)

    assert len(tree.get_active_nodes()) == 0


def test_toggle_joint_segments():
    tree = ToggleSegmentTree(0, 20)

    tree.toggle(10, 15)
    tree.toggle(12, 18)
    stack_interval_active = tree.get_stack_of_active_intervals()
    assert stack_interval_active == [10, 12, 15, 18]


def test_toggle_disjoint_segments():
    tree = ToggleSegmentTree(0, 20)

    tree.toggle(5, 10)
    tree.toggle(15, 19)
    stack_interval_active = tree.get_stack_of_active_intervals()
    assert stack_interval_active == [5, 10, 15, 19]
